sktime/decomposition/vampnet.py

```diff
@@ -0,0 +1,662 @@
+
+    if remove_mean:
+        x = x - x.mean(dim=0, keepdim=True)
+        y = y - y.mean(dim=0, keepdim=True)
+
+    # Calculate the cross-covariance
+    y_t = y.transpose(0, 1)
+    x_t = x.transpose(0, 1)
+    cov_01 = 1 / (batch_size - 1) * torch.matmul(x_t, y)
+    # Calculate the auto-correlations
+    cov_00 = 1 / (batch_size - 1) * torch.matmul(x_t, x)
+    cov_11 = 1 / (batch_size - 1) * torch.matmul(y_t, y)
+
+    return cov_00, cov_01, cov_11
+
+
+valid_score_methods = ('VAMP1', 'VAMP2', 'VAMPE')
+
+
+def score(data: torch.Tensor, data_lagged: torch.Tensor, method='VAMP2', epsilon: float = 1e-6, mode='trunc'):
+    r"""Computes the VAMP score based on data and corresponding time-shifted data.
+
+    Parameters
+    ----------
+    data : torch.Tensor
+        (N, d)-dimensional torch tensor
+    data_lagged : torch.Tensor
+        (N, k)-dimensional torch tensor
+    method : str, default='VAMP2'
+        The scoring method. See :meth:`score <sktime.decomposition.CovarianceKoopmanModel.score>` for details.
+    epsilon : float, default=1e-6
+        Cutoff parameter for small eigenvalues, alternatively regularization parameter.
+    mode : str, default='trunc'
+        Regularization mode for Hermetian inverse. See :meth:`sym_inverse`.
+
+    Returns
+    -------
+    score : torch.Tensor
+        The score.
+    """
+    if method not in valid_score_methods:
+        raise ValueError(f"Invalid method '{method}', supported are {valid_score_methods}")
+    assert data.shape == data_lagged.shape
+
+    if method == 'VAMP1':
+        koopman = koopman_matrix(data, data_lagged, epsilon=epsilon, mode=mode)
+        vamp_score = torch.norm(koopman, p='nuc')
+    elif method == 'VAMP2':
+        koopman = koopman_matrix(data, data_lagged, epsilon=epsilon, mode=mode)
+        vamp_score = torch.pow(torch.norm(koopman, p='fro'), 2)
+    elif method == 'VAMPE':
+        c00, c0t, ctt = covariances(data, data_lagged, remove_mean=True)
+        c00_sqrt_inv = sym_inverse(c00, epsilon=epsilon, ret_sqrt=True, mode=mode)
+        ctt_sqrt_inv = sym_inverse(ctt, epsilon=epsilon, ret_sqrt=True, mode=mode)
+        koopman = torch.chain_matmul(c00_sqrt_inv, c0t, ctt_sqrt_inv).t()
+
+        u, s, v = torch.svd(koopman)
+        mask = s > epsilon
+
+        u = torch.mm(c00_sqrt_inv, u[:, mask])
+        v = torch.mm(ctt_sqrt_inv, v[:, mask])
+        s = s[mask]
+
+        u_t = u.t()
+        v_t = v.t()
+        s = torch.diag(s)
+
+        vamp_score = torch.trace(
+            2. * torch.chain_matmul(s, u_t, c0t, v)
+            - torch.chain_matmul(s, u_t, c00, u, s, v_t, ctt, v)
+        )
+    else:
+        raise RuntimeError("This should have been caught earlier.")
+    return 1 + vamp_score
+
+
+def loss(data: torch.Tensor, data_lagged: torch.Tensor, method='VAMP2', epsilon: float = 1e-6, mode: str = 'trunc'):
+    r"""Loss function that can be used to train VAMPNets. It evaluates as :math:`1-\mathrm{score}`. The score
```

Comment:
why does it evaluate as 1 - score if it's implemented as -1 * score?

Reply:
that's a typo

Reply:
fixed